Added IsURL(), ToURL(), PushURL() and 2 variants of CheckURL() to dmsdk.

Gert-Jan opened the pull request:
Allows native extensions to have better control over URLs in the lua state.
Closes https://github.com/defold/defold/issues/7220
### Technical changes
* Moved `IsURL()` in `script_msg.cpp` to the other URL related functions. Changed the `IsURL()` implementation to do the same as other [`IsXxx()`](https://github.com/defold/defold/blob/027491c8d5dbe0b592a7d41f348056f902c78113/engine/script/src/script_vmath.cpp#L188) implentations.
* As it stands the URL related implementations are split in [script_msg.cpp](https://github.com/defold/defold/blob/a562970c7fea4e059faf1f57ca704625fa1090ab/engine/script/src/script_msg.cpp#L582) and [script.cpp](https://github.com/defold/defold/blob/a562970c7fea4e059faf1f57ca704625fa1090ab/engine/script/src/script.cpp#L848). For now I put the implementation for CheckURL() in script.cpp because it's so closely related to it. Maybe a better place would be script_msg.cpp because there already is [different variant](https://github.com/defold/defold/blob/a562970c7fea4e059faf1f57ca704625fa1090ab/engine/script/src/script_msg.cpp#L590) of CheckURL().
* The specific use case for this change was getting an absolute URL from a relative URL string, when calling from Lua. The rest (IsURL(), ToURL(), PushURL() and other CheckURL()) of this PR escalated from there. Example of getting an absolute URL:
```
// called from lua
dmMessage::URL sender;
dmMessage::URL absoluteUrl;
if (dmScript::CheckURL(L, &sender)) {
  // lua stack 1: "object#component"
  dmScript::ResolveURL(L, 1, &absoluteUrl, &sender);
}
```

Gert-Jan committed:
Added CheckURL() to dmsdk. Which is a wrapper for ToURL().

JCash commented on engine/script/src/dmsdk/script/script.h:
```diff
@@ -640,6 +640,15 @@ namespace dmScript
+     * @param out_url [type:dmMessage::URL*] where to store the result
+     * @return result [type:bool] true if successful
+     */
+    bool CheckURL(lua_State* L, dmMessage::URL* out_url);
```
Some api inconsistencies:
* The `CheckXxx` api should return the value type, which is (dmMessage::URL*) in this case.
* It should accept the `index` of the argument on the Lua stack
* Each type `Xxx`should have `IsXxx()`, `CheckXxx()`, `ToXxx()` and `PushXxx()`,

Note that we already have these functions, and they should only require to be moved, and update the documentation.

JCash replied:
Right, yeah, it is a good use case for getting the url of the current script.

> how does the user know when CheckURL failed?

The "CheckXxx" should always throw an error if they fail.
If we want to implement something more graceful, that's what we use the IsXxx/ToXxx for.

Gert-Jan committed:
Added IsURL, ToURL, PushURL and CheckURL(lua_State* L, int index)j.

Moved CheckURL(lua_State* L, dmMessage::URL* out_url) up so it's close to the other url getters.
Implemented ToURL().
Moved down the IsURL() implementation so it's near the other URL functions.

Gert-Jan committed:
A more standard implementation for IsXxx().

JCash commented on engine/script/src/script.cpp:
```diff
@@ -872,6 +872,14 @@ namespace dmScript
         return false;
     }
 
+    bool CheckURL(lua_State* L, dmMessage::URL* out_url) {
```
style: the curly brace should be on a new line.

JCash commented on engine/script/src/script.cpp:
```diff
@@ -872,6 +872,14 @@ namespace dmScript
+    bool CheckURL(lua_State* L, dmMessage::URL* out_url) {
+        bool result = GetURL(L, out_url);
+        if (!result) {
+            luaL_error(L, "no URL could be found in the current script environment");
```
The recommendation from Lua is to always do a return: "return luaL_error()"

Also type: "no" -> "No"

JCash reviewed (changes requested):
lgtm:
Minor style changes needed

Gert-Jan committed:
Small style tweaks.

AGulev commented:
It also fixes this issue https://github.com/defold/defold/issues/7220

Please, add it into your PR description (see explanation here: https://github.com/defold/defold/pull/10767#issuecomment-2939849009)
